client/extensions/woocommerce/app/settings/email/mailchimp/setup-mailchimp.js

```diff
@@ -0,0 +1,315 @@
+			( this.state.step === CAMPAIGN_DEFAULTS_STEP ) ) {
+			this.setState( { step: NEWSLETTER_SETTINGS_STEP } );
+		} else if ( ( nextProps.settings.active_tab === STORE_SYNC ) &&
+			( this.state.step === NEWSLETTER_SETTINGS_STEP ) ) {
+			nextProps.onClose();
+		}
+	}
+
+	prepareDefaultValues( settings ) {
+		const newSettings = Object.assign( {}, settings );
+		newSettings.campaign_from_name = settings.campaign_from_name || settings.store_name || '';
+		newSettings.campaign_from_email = settings.campaign_from_email || settings.admin_email || '';
+		newSettings.campaign_subject = settings.campaign_subject || settings.store_name || '';
+		newSettings.store_locale = settings.store_locale || 'en';
+		newSettings.campaign_language = settings.campaign_language || settings.store_locale;
+		newSettings.campaign_permission_reminder = settings.campaign_permission_reminder ||
+			'You were subscribed to the newsletter from ' + settings.store_name;
+		return newSettings;
+	}
+
+	onClose = () => {
+		this.props.onClose();
+	}
+
+	getStoreSettings = () => {
+		// clear this and pass only what is required.
+		const { address, currency } = this.props;
+
+		const settings = pick( this.state.settings, storSettingsRequriredFields );
+		settings.store_city = address.city;
+		settings.store_street = address.street + ' ' + address.street2;
+		settings.store_state = address.state;
+		settings.store_country = address.country;
+		settings.store_postal_code = address.postcode;
+		settings.store_currency_code = currency;
+		return settings;
+	}
+
+	getCampaingDefaultsSettings = () => {
+		return pick( this.state.settings, campaignDefaultsRequiredFields );
+	}
+
+	hasEmptyValues = ( data ) => {
+		return some( data, isEmpty );
+	}
+
+	areStoreSettingsValid = ( settings ) => {
+		const hasAllKeys = storSettingsRequriredFields.every( key => key in settings );
+		if ( ! hasAllKeys ) {
+			return false;
+		}
+		const hasEmptyValues = this.hasEmptyValues( settings );
+		if ( hasEmptyValues ) {
+			return false;
+		}
+		if ( settings.store_phone.length <= 6 ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	areCampaignSettingsValid = ( settings ) => {
+		const hasAllKeys = campaignDefaultsRequiredFields.every( key => key in settings );
+		if ( ! hasAllKeys ) {
+			return false;
+		}
+		return true;
+	}
+
+	next = () => {
+		// Don't send data if it is not compleate or it is the same as the one we already had from
+		// the endpoint.
+		if ( this.state.step === KEY_INPUT_STEP ) {
+			this.props.submitMailChimpApiKey( this.props.siteId, this.state.api_key_input );
+			return;
+		} else if ( this.state.step === STORE_INFO_STEP ) {
+			const settings = this.getStoreSettings();
+			const validSettings = this.areStoreSettingsValid( settings );
+			if ( ! validSettings ) {
+				this.setState( { settings_values_missing: true } );
+				return;
+			}
+			this.setState( { settings_values_missing: false } );
+
+			this.props.submitMailchimpStoreInfo( this.props.siteId, settings );
```

Comment:
should this be called before the setState call? Same for all the other if clauses below.